text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+`std::module_path` and expands to an expression of type `&'static str`.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently macros for getting the source file name, `std::file`, and
+the line number, `std::line`. To complete this pair, a macro to for getting
+the current function name would be useful. All three macros are useful for
+e.g. generating log output with more context about where something was
+happening. The file and line is good for being able to look up the exact
+location, however the fully qualified function name gives additional context
+without even having to look into the source code.
+
+# Detailed design
+[design]: #detailed-design
+
+This feature would have an equivalent design and implementation to
+`std::file`. It would return a fully qualified function name but the actual
+string returned is implementation defined. The string is meant to be used by
+the user to get an idea about the context of the code in e.g. debug output,
+and should be meaningful for a human, but is not meant to be analyzed by a
+program.
+
+For example the following scheme could be used:
+
+ - Plain functions use their name, e.g. `some_function` for
+
+````
+    fn some_function(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Generic functions use their name, followed by the generic arguments, e.g.
+   `some_function<T, U>` for
+
+````
+    fn some_function<T, U>(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Functions inside non-generic `impl` blocks for structs would use the name of the
+   struct, followed by `::` and the function name, e.g.
+   `SomeStruct::some_function` for
+
+````
+    struct SomeStruct {...}
+
+    impl SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside generic `impl` blocks for structs would use the name of the
+   struct, followed by the generic parameters as given by the `impl` block,
+   followed by `::` and the function name, e.g.
+   `SomeStruct<i32>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl SomeStruct<i32> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Similar for generic `impl` blocks, e.g. `SomeStruct<T>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl<T> SomeStruct<T> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside trait `impl` blocks would use, enclosed in angle brackets,
+   the name of the type, followed by `as`, followed by the trait name, and
+   then `::` followed by the name of the function, e.g.
+   `<SomeStruct as SomeTrait>::some_function` for
+
+````
+    impl SomeTrait for SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Closures are using the name of the surrounding function, followed by `::`
+   and the string `<closure/definition_line_number>`, where `definition_line_number` is the number
```

Comment:
> So rather than having to add line!() again, why not just start with the information to begin with?

We already have an intrinsic for similar (debugging) purposes (cough… `::core::intrinsics::type_name` cough…) which in my experience outputs too much information than desired (i.e. includes lifetimes and full path for types from external crates) resulting in the end user (cough… rustc… cough) using hacky means to decompose output into the necessary format, rather than composing it from small parts (which is what would make the most sense in the first place).

What I’m trying to say, is that even provided the same debugging use-case the desired output is likely different.

---

For all its worth, both g++ and clang++ code below prints just `operator()`, so no prior art. It does point to the fact that we could output some form of UFCS (i.e. `<_ as Fn(args) -> R>::call()`).

```
int main() {
  auto c = [&]() {
    std::cout << __func__ << std::endl;
    return 0;
  };
  return c();
}
```